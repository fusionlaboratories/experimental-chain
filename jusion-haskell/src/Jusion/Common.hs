module Jusion.Common where

type Address = Int
type Amount = Int
type Height = Int