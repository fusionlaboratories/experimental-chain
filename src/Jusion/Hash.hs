{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE ImportQualifiedPost #-}

module Jusion.Hash (Hash, hash) where

import Data.Hashable qualified as H
import GHC.Generics

-- NOTE, we are not using a cryptographically safe Hash
newtype Hash a = Hash Int
    deriving (Eq, Show, Generic)

instance (H.Hashable a) => H.Hashable (Hash a)

hash :: (H.Hashable a) => a -> Hash a
hash = Hash . H.hash