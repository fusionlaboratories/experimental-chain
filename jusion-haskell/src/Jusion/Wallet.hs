{-# LANGUAGE TemplateHaskell #-}
module Jusion.Wallet where

import Control.Lens.TH

import Jusion.Common

data Wallet tx = Wallet
    { _address :: Address
    , _balance :: Amount
    , _transactions :: [tx]
    }
    deriving (Eq, Show)

makeLenses ''Wallet
