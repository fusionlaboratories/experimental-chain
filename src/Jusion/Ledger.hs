module Jusion.Ledger where

import Jusion.Common

import Data.Map.Strict as Map

newtype Ledger = Ledger
    { _byAddress :: Map.Map Address Amount
    }
    deriving (Eq, Show)