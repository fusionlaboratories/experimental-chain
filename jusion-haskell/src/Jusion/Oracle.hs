module Jusion.Oracle where

import Control.Monad.Reader

import Data.Map.Strict qualified as Map
import Data.IntSet qualified as IntSet

import Jusion.Common
import Jusion.Hash ( Hash )

-- TODO: What if this was just a bloom filter based on block hashes?
-- many APIs and services actually assume that a block hash is unique...  What
-- is the possibility of collision?
data Blocks block = Blocks
    { _byHeight :: Map.Map Height (Hash block)
    , _hashes :: IntSet.IntSet
    }
newtype L1 block = L1 { blocks :: Map.Map Height (Hash block) }
    deriving (Eq, Show)

-- Here are the operations directly on the L1 Oracle
observeBlock :: MonadReader (L1 block) m => Height -> Hash block -> m ()
observeBlock = undefined

