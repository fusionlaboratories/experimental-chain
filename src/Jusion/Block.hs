{-# LANGUAGE DeriveGeneric, TemplateHaskell #-}
module Jusion.Block where

import Control.Lens.TH

import Data.Hashable

import GHC.Generics

import Jusion.Common
import Jusion.Hash

data Block tx = Block
    { _parent :: Hash (Block tx)
    , _height :: Integer
    , _transactions :: [tx]
    }
    deriving (Eq, Show, Generic)

makeLenses ''Block

instance (Hashable tx) => Hashable (Block tx)
