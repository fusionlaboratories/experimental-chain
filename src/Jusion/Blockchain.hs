{-# LANGUAGE TemplateHaskell #-}
module Jusion.Blockchain where

import Data.Map.Strict as Map

import Control.Lens.TH

import Jusion.Common
import Jusion.Hash

data Blockchain block = Blockchain
    { _byHeight :: Map.Map Height block
    , _byHash :: Map.Map (Hash block) block
    }
    deriving (Eq, Show)

makeLenses ''Blockchain