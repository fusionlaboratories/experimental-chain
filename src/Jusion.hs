{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE ImportQualifiedPost #-}

module Jusion where

import Data.Hashable (Hashable)
import Data.Map.Strict qualified as Map

import GHC.Generics (Generic)

import Jusion.Hash
import Jusion.Common
import Jusion.Wallet qualified as Wallet
import Jusion.Wallet hiding (Wallet)

-- TODO: Implement a simple model of Jusion

-- L1 Oracle, which contains
-- A Map of Blocks


-- Wallet
type Wallet = Wallet.Wallet Transaction

-- NOTE: Does not fully model the origin
data Block = Block
    { parent :: Hash Block
    , height :: Integer
    , transactions :: [Transaction]
    }
    deriving (Eq, Show, Generic)

instance Hashable Block

data Transaction = Transaction
    { from :: Address
    , to :: Address
    , amount :: Integer
    , block :: Block
    }
    deriving (Eq, Show, Generic)

instance Hashable Transaction

newtype Ledger = Ledger
    { byAddress :: Map.Map Address Amount
    }
    deriving (Eq, Show)

data Blockchain = Blockchain
    { byHeight :: Map.Map Height Block
    , byHash :: Map.Map (Hash Block) Block
    }
    deriving (Eq, Show)

data TransactionLog = TransactionLog
    { byHeight :: Map.Map Height Transaction
    , byHash :: Map.Map (Hash Transaction) Transaction
    }
    deriving (Eq, Show)

data Network = Network
    { ledger :: Ledger
    , blockchain :: Blockchain
    , transactionLog :: TransactionLog
    }
    deriving (Eq, Show)
