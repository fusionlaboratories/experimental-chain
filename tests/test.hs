module Main where

import Test.Tasty
import Test.Tasty.HUnit ( testCase, (@?=) )

import Jusion

main :: IO ()
main = defaultMain tests

tests :: TestTree
tests =
    testGroup "finite fields" [
        testCase "f17 wraps around" $ (0 :: F17) @?= (17 :: F17)
    ]