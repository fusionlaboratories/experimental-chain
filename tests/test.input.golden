{"operand_stack":["0","18446744069414584321"]}